File: javascript/tests/e2e.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert';
import { createProvider } from '../src/index.js';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

// Load .env (Node doesn't auto-load it)
try {
  const envPath = join(dirname(fileURLToPath(import.meta.url)), '../.env');
  for (const line of readFileSync(envPath, 'utf8').split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eq = trimmed.indexOf('=');
    if (eq < 0) continue;
    const k = trimmed.slice(0, eq).trim();
    const v = trimmed.slice(eq + 1).trim();
    if (!process.env[k]) process.env[k] = v;
  }
} catch { /* .env optional */ }

const UA_POOL = [
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15",
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0",
];

const RESEND_API_KEY = process.env.RESEND_API_KEY || '';

/**
 * Sends a test email to the specified address via Resend API.
 * @param {string} to - Recipient email address
 * @returns {Promise<boolean>}
 */
async function sendTestEmail(to) {
  const delays = [1000, 3000, 5000];
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const resp = await fetch('https://api.resend.com/emails', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${RESEND_API_KEY}`,
          'User-Agent': UA_POOL[Math.floor(Math.random() * UA_POOL.length)],
        },
        body: JSON.stringify({
          from: 'onboarding@rokupusu.web.id',
          to,
          subject: 'TempMail E2E Test',
          html: '<p>E2E test email from TempMail wrapper</p>',
        }),
      });
      if (resp.ok) return true;
      if (attempt < 2) {
        await new Promise(r => setTimeout(r, delays[attempt]));
        continue;
      }
      return false;
    } catch {
      if (attempt < 2) {
        await new Promise(r => setTimeout(r, delays[attempt]));
      }
      continue;
    }
  }
  return false;
}

/**
 * Helper to run provider E2E test.
 * @param {string} providerName
 * @param {object} [config]
 */
async function testProvider(providerName, config) {
  const provider = createProvider(providerName, config);
  assert.ok(provider, `${providerName}: provider should be created`);

  // Step 1: Generate email
  let email;
  try {
    email = await provider.generateEmail();
    console.log(`  ${providerName}: generated ${email}`);
  } catch (err) {
    console.warn(`  ${providerName}: generateEmail failed (skipping): ${err.message}`);
    return; // Skip rest of test
  }

  assert.ok(email.includes('@'), `${providerName}: email should contain @`);

  // Step 2: Send test email and wait for delivery
  let emailSent = false;
  try {
    emailSent = await sendTestEmail(email);
    if (emailSent) {
      console.log(`  ${providerName}: test email sent to ${email}`);
      await new Promise(r => setTimeout(r, 4000));
    } else {
      console.warn(`  ${providerName}: sendTestEmail failed (skipping inbox assertion)`);
    }
  } catch (err) {
    console.warn(`  ${providerName}: sendTestEmail error: ${err.message}`);
  }

  // Step 3: Get inbox
  let inbox = [];
  try {
    inbox = await provider.getInbox(email);
    console.log(`  ${providerName}: inbox has ${inbox.length} messages`);
  } catch (err) {
    console.warn(`  ${providerName}: getInbox failed: ${err.message}`);
  }

  assert.ok(Array.isArray(inbox), `${providerName}: inbox should be array`);

  // If send succeeded, warn if empty
  if (emailSent && inbox.length === 0) {
    console.warn(`  ${providerName}: WARNING: test email was sent but inbox is empty (delivery may be slow)`);
  }

  // Step 3: If inbox has messages, read the first one
  if (inbox.length > 0) {
    const msg = inbox[0];
    assert.ok(msg.id, `${providerName}: message should have id`);
    assert.ok(msg.sender, `${providerName}: message should have sender`);

    try {
      // Use format "id:email" for providers that need it
      const messageId = `${msg.id}:${email}`;
      const detail = await provider.readMessage(messageId);
      console.log(`  ${providerName}: read message ${msg.id}`);

      assert.ok(detail.id, `${providerName}: detail should have id`);
      assert.ok(typeof detail.bodyText === 'string', `${providerName}: detail should have bodyText`);
    } catch (err) {
      console.warn(`  ${providerName}: readMessage failed: ${err.message}`);
    }
  } else {
    console.log(`  ${providerName}: no messages to read (expected for fresh address)`);
  }

  // Step 4: Delete email
  try {
    const deleted = await provider.deleteEmail(email);
    console.log(`  ${providerName}: deleteEmail returned ${deleted}`);
    assert.ok(typeof deleted === 'boolean', `${providerName}: deleteEmail should return boolean`);
  } catch (err) {
    console.warn(`  ${providerName}: deleteEmail failed: ${err.message}`);
  }
}

describe('E2E Provider Tests', () => {
  describe('mailtm', () => {
    it('should complete full flow', async () => {
      await testProvider('mailtm');
    });
  });

  describe('guerrillamail', () => {
    it('should complete full flow', async () => {
      await testProvider('guerrillamail');
    });
  });

  describe('yopmail', () => {
    it('should complete full flow', async () => {
      await testProvider('yopmail');
    });
  });

  describe('dropmail', () => {
    it('should complete full flow', async () => {
      await testProvider('dropmail');
    });
  });

  describe('1secemail', () => {
    it('should complete full flow', async () => {
      await testProvider('1secemail');
    });
  });

  describe('ncaori', () => {
    it('should complete full flow', async () => {
      await testProvider('ncaori');
    });
  });

  describe('zoromail', () => {
    it('should complete full flow', async () => {
      await testProvider('zoromail');
    });
  });

  describe('tempmail.lol', () => {
    it('should complete full flow', async () => {
      await testProvider('tempmail.lol');
    });
  });

  describe('tempmailc', () => {
    it('should complete full flow', async () => {
      await testProvider('tempmailc');
    });
  });

  describe('temp-mail.io', () => {
    it('should complete full flow', async () => {
      await testProvider('temp-mail.io');
    });
  });

  describe('tempmail.plus', () => {
    it('should complete full flow', async () => {
      await testProvider('tempmail.plus');
    });
  });

  describe('emailfake', () => {
    it('should complete full flow', async () => {
      await testProvider('emailfake');
    });
  });

  describe('generator.email', () => {
    it('should complete full flow', async () => {
      await testProvider('generator.email');
    });
  });

  describe('mailnesia', () => {
    it('should complete full flow', async () => {
      await testProvider('mailnesia');
    });
  });

  describe('10minutemail', () => {
    it('should complete full flow', async () => {
      await testProvider('10minutemail');
    });
  });

  describe('email-temp', () => {
    it('should complete full flow', async () => {
      await testProvider('email-temp');
    });
  });
});

describe('Factory Tests', () => {
  it('should create yopmail provider with alternate name', () => {
    const provider = createProvider('yop');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'yopmail');
  });

  it('should create dropmail provider with alternate name', () => {
    const provider = createProvider('dropmail.me');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'dropmail');
  });

  it('should create 1secemail provider', () => {
    const provider = createProvider('1secemail');
    assert.ok(provider);
    assert.strictEqual(provider.name, '1secemail');
  });

  it('should create ncaori provider', () => {
    const provider = createProvider('ncaori');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'ncaori');
  });

  it('should create ncaori provider with alternate name', () => {
    const provider = createProvider('nca.my.id');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'ncaori');
  });

  it('should create zoromail provider', () => {
    const provider = createProvider('zoromail');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'zoromail');
  });

  it('should create tempmail.lol provider', () => {
    const provider = createProvider('tempmail.lol');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'tempmail.lol');
  });

  it('should create tempmailc provider', () => {
    const provider = createProvider('tempmailc');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'tempmailc');
  });

  it('should create temp-mail.io provider', () => {
    const provider = createProvider('temp-mail.io');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'temp-mail.io');
  });

  it('should create tempmail.plus provider', () => {
    const provider = createProvider('tempmail.plus');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'tempmail.plus');
  });

  it('should create emailfake provider', () => {
    const provider = createProvider('emailfake');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'emailfake');
  });

  it('should create generator.email provider', () => {
    const provider = createProvider('generator.email');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'generator.email');
  });

  it('should create mailnesia provider', () => {
    const provider = createProvider('mailnesia');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'mailnesia');
  });

  it('should create 10minutemail provider', () => {
    const provider = createProvider('10minutemail');
    assert.ok(provider);
    assert.strictEqual(provider.name, '10minutemail');
  });

  it('should create email-temp provider', () => {
    const provider = createProvider('email-temp');
    assert.ok(provider);
    assert.strictEqual(provider.name, 'email-temp');
  });

  it('should throw on unknown provider', () => {
    assert.throws(() => createProvider('unknown'), /Unknown provider/);
  });
});

File: java/src/test/java/com/tempmail/E2ETest.java
package com.tempmail;

import com.tempmail.errors.TempMailException;
import com.tempmail.models.Message;
import com.tempmail.models.MessageDetail;
import com.tempmail.providers.*;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;

/**
 * Simple E2E test runner — no JUnit needed.
 * Run: mvn compile exec:java -Dexec.mainClass="com.tempmail.E2ETest"
 * Or:  javac + java directly
 */
public class E2ETest {

    private static int passed = 0;
    private static int failed = 0;
    private static int skipped = 0;

    public static void main(String[] args) {
        // Load .env file (Java doesn't auto-load it)
        java.io.File envFile = new java.io.File(".env");
        if (!envFile.exists()) envFile = new java.io.File("../.env");
        if (envFile.exists()) {
            try (java.util.Scanner sc = new java.util.Scanner(envFile)) {
                while (sc.hasNextLine()) {
                    String line = sc.nextLine().trim();
                    if (line.isEmpty() || line.startsWith("#")) continue;
                    int eq = line.indexOf('=');
                    if (eq < 0) continue;
                    String k = line.substring(0, eq).trim();
                    String v = line.substring(eq + 1).trim();
                    if (System.getenv(k) == null) {
                        // ponytail: ProcessBuilder env injection — set in System props as fallback
                        System.setProperty("env." + k, v);
                    }
                }
            } catch (Exception ignored) {}
        }
        System.out.println("=== TempMail E2E Tests ===\n");

        testFactory();
        testProvider("mail_tm", new MailTmProvider());
        testProvider("guerrilla_mail", new GuerrillaMailProvider());
        testProvider("yopmail", new YOPmailProvider());
        testProvider("dropmail", new DropmailProvider());
        testProvider("1secemail", new OneSecEmailProvider());
        testProvider("ncaori", new NcaoriMailProvider());
        testProvider("zoromail", new ZoromailProvider());
        testProvider("tempmail_lol", new TempmailLolProvider());
        testProvider("tempmailc", new TempmailcProvider());
        testProvider("temp_mail_io", new TempMailIoProvider());
        testProvider("tempmail_plus", new TempmailPlusProvider());
        testProvider("emailfake", new EmailfakeProvider());
        testProvider("generator_email", new GeneratorEmailProvider());
        testProvider("mailnesia", new MailnesiaProvider());
        testProvider("tenminutemail", new TenMinuteMailProvider());
        testProvider("email_temp", new EmailTempProvider());

        System.out.println("\n=== Results ===");
        System.out.println("Passed: " + passed);
        System.out.println("Failed: " + failed);
        System.out.println("Skipped: " + skipped);
        System.out.println("Total:  " + (passed + failed + skipped));

        if (failed > 0) {
            System.exit(1);
        }
    }

    private static void testFactory() {
        System.out.println("[factory] Testing TempMailFactory.create()...");
        try {
            assertCondition("yopmail factory",
                    TempMailFactory.create("yopmail") instanceof YOPmailProvider);
            assertCondition("dropmail factory",
                    TempMailFactory.create("dropmail") instanceof DropmailProvider);
            assertCondition("mail_tm factory",
                    TempMailFactory.create("mail_tm") instanceof MailTmProvider);
            assertCondition("guerrilla_mail factory",
                    TempMailFactory.create("guerrilla_mail") instanceof GuerrillaMailProvider);
            assertCondition("ncaori factory",
                    TempMailFactory.create("ncaori") instanceof NcaoriMailProvider);
            assertCondition("zoromail factory",
                    TempMailFactory.create("zoromail") instanceof ZoromailProvider);
            assertCondition("tempmail_lol factory",
                    TempMailFactory.create("tempmail.lol") instanceof TempmailLolProvider);
            assertCondition("tempmailc factory",
                    TempMailFactory.create("tempmailc") instanceof TempmailcProvider);
            assertCondition("temp_mail_io factory",
                    TempMailFactory.create("temp-mail.io") instanceof TempMailIoProvider);
            assertCondition("tempmail_plus factory",
                    TempMailFactory.create("tempmail.plus") instanceof TempmailPlusProvider);
            assertCondition("emailfake factory",
                    TempMailFactory.create("emailfake") instanceof EmailfakeProvider);
            assertCondition("generator_email factory",
                    TempMailFactory.create("generator.email") instanceof GeneratorEmailProvider);
            assertCondition("mailnesia factory",
                    TempMailFactory.create("mailnesia") instanceof MailnesiaProvider);
            assertCondition("tenminutemail factory",
                    TempMailFactory.create("10minutemail") instanceof TenMinuteMailProvider);
            assertCondition("email_temp factory",
                    TempMailFactory.create("email-temp") instanceof EmailTempProvider);
        } catch (Exception e) {
            fail("factory", e);
        }
    }

    private static void testProvider(String name, TempMailProvider provider) {
        System.out.println("\n[" + name + "] Starting E2E test...");
        try {
            // Step 1: Generate email (with retry for mail_tm)
            String email;
            if ("mail_tm".equals(name)) {
                email = generateWithRetry(name, provider);
                if (email == null) {
                    System.out.println("SKIP: mail.tm failed after 3 retry attempts");
                    skipped += 5;
                    return;
                }
            } else {
                email = provider.generateEmail();
            }
            assertCondition(name + " generateEmail", email != null && email.contains("@"));
            System.out.println("[" + name + "] Email: " + email);

            // Step 1b: Send a test email to this address
            boolean sent = sendTestEmail(email);
            if (sent) {
                System.out.println("[" + name + "] Test email sent, waiting 4s for delivery...");
                try { Thread.sleep(4000); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); }
            } else {
                System.out.println("[" + name + "] WARN: sendTestEmail failed, skipping inbox delivery assertion");
            }

            // Step 2: Get inbox
            List<Message> inbox = provider.getInbox(email);
            assertCondition(name + " getInbox", inbox != null);
            System.out.println("[" + name + "] Inbox size: " + inbox.size());
            if (sent && (inbox == null || inbox.isEmpty())) {
                System.out.println("[" + name + "] WARN: sent test email but inbox still empty (delivery delay?)");
            }

            // Step 3: Read message if available
            if (!inbox.isEmpty()) {
                Message first = inbox.get(0);
                assertCondition(name + " message id", first.getId() != null);
                System.out.println("[" + name + "] Reading message: " + first.getId());

                MessageDetail detail = provider.readMessage(first.getId());
                assertCondition(name + " readMessage", detail != null);
                System.out.println("[" + name + "] Subject: " + detail.getSubject());
            } else {
                System.out.println("[" + name + "] No messages (expected for new addresses)");
            }

            // Step 4: Delete
            try {
                boolean deleted = provider.deleteEmail(email);
                if (!deleted) System.out.println("[" + name + "] WARN: deleteEmail returned false (may be unsupported)");
                else passed++;
            } catch (Exception e) {
                System.out.println("[" + name + "] WARN: deleteEmail failed: " + e.getMessage() + " (may be unsupported)");
            }
            System.out.println("[" + name + "] E2E PASSED");

        } catch (TempMailException e) {
            // Handle known provider-side failures gracefully
            String msg = e.getMessage() != null ? e.getMessage() : "";
            int code = e.getStatusCode();
            if ((code == 402) || (code == 400 && msg.contains("yopmail"))
                    || msg.contains("captcha_required") || msg.contains("HTTP 400")
                    || msg.contains("HTTP 402")) {
                System.out.println("SKIP: " + name + " provider unavailable (" + code + "): " + msg.substring(0, Math.min(80, msg.length())));
                skipped += 5;
                return;
            }
            // Other TempMailException: fail
            fail(name, e);
        } catch (Exception e) {
            fail(name, e);
        }
    }

    /**
     * Retry generateEmail() with backoff: 1s, 3s, 5s.
     * Returns null if all 3 attempts fail.
     */
    private static String generateWithRetry(String name, TempMailProvider provider) {
        int[] delays = {1000, 3000, 5000};
        for (int attempt = 1; attempt <= 3; attempt++) {
            try {
                return provider.generateEmail();
            } catch (Exception e) {
                System.out.println("[" + name + "] generateEmail attempt " + attempt + "/3 failed: " + e.getMessage());
                if (attempt < 3) {
                    try {
                        Thread.sleep(delays[attempt - 1]);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        System.out.println("[" + name + "] Interrupted during retry wait");
                        return null;
                    }
                }
            }
        }
        return null;
    }

    private static final String[] UA_POOL = {
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0",
    };

    private static String randomUA() {
        return UA_POOL[(int)(Math.random() * UA_POOL.length)];
    }

    private static String resendApiKey() {
        String k = System.getenv("RESEND_API_KEY");
        if (k != null && !k.isEmpty()) return k;
        // ponytail: fallback to .env loaded via System properties
        k = System.getProperty("env.RESEND_API_KEY");
        return (k != null) ? k : "";
    }

    private static boolean sendTestEmail(String to) {
        int[] delays = {1000, 3000, 5000};
        for (int attempt = 0; attempt < 3; attempt++) {
            try {
                HttpClient client = HttpClient.newBuilder()
                        .connectTimeout(Duration.ofSeconds(10))
                        .build();
                String escaped = to.replace("\"", "\\\"");
                String body = "{\"from\":\"onboarding@rokupusu.web.id\",\"to\":\"" + escaped + "\",\"subject\":\"TempMail E2E Test\",\"html\":\"<p>E2E test email from TempMail wrapper</p>\"}";
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(java.net.URI.create("https://api.resend.com/emails"))
                        .header("Content-Type", "application/json")
                        .header("Authorization", "Bearer " + resendApiKey())
                        .header("User-Agent", randomUA())
                        .timeout(Duration.ofSeconds(10))
                        .POST(HttpRequest.BodyPublishers.ofString(body))
                        .build();
                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
                System.out.println("  [sendTestEmail] Status: " + response.statusCode());
                if (response.statusCode() == 200) return true;
                if (attempt < 2) {
                    try { Thread.sleep(delays[attempt]); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); return false; }
                    continue;
                }
                return false;
            } catch (Exception e) {
                System.out.println("  [sendTestEmail] WARN attempt " + (attempt + 1) + "/3: " + e.getMessage());
                if (attempt < 2) {
                    try { Thread.sleep(delays[attempt]); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); return false; }
                    continue;
                }
                return false;
            }
        }
        return false;
    }

    private static void assertCondition(String label, boolean condition) {
        if (condition) {
            passed++;
        } else {
            failed++;
            System.err.println("  FAIL: " + label);
        }
    }

    private static void fail(String name, Exception e) {
        failed++;
        System.err.println("[" + name + "] FAILED: " + e.getMessage());
        e.printStackTrace(System.err);
    }
}
